app/channel.go

```diff
@@ -722,38 +722,107 @@ func (a *App) GetTeamSchemeRolesForChannel(teamId string) (string, string, strin
+	inheritedMemberRole, err := a.GetRoleByName(inheritedMemberRoleName)
+	if err != nil {
+		return nil, err
+	}
+
+	inheritedGuestRole, err := a.GetRoleByName(inheritedGuestRoleName)
 	if err != nil {
 		return nil, err
 	}
 
-	memberPermissions := GetChannelModeratedPermissions(memberRole)
-	guestPermissions := GetChannelModeratedPermissions(guestRole)
+	return buildChannelModerations(memberRole, guestRole, inheritedMemberRole, inheritedGuestRole), nil
+}
 
-	inhertedGuestRoleName, inheritedMemberRoleName, _, _ := a.GetTeamSchemeRolesForChannel(channel.TeamId)
+func (a *App) PatchChannelModerationsForChannel(channel *model.Channel, channelModerationsPatch []*model.ChannelModerationPatch) ([]*model.ChannelModeration, *model.AppError) {
+	inheritedGuestRoleName, inheritedMemberRoleName, _, _ := a.GetTeamSchemeRolesForChannel(channel.TeamId)
 	inheritedMemberRole, err := a.GetRoleByName(inheritedMemberRoleName)
 	if err != nil {
 		return nil, err
 	}
 
-	inhertedGuestRole, err := a.GetRoleByName(inhertedGuestRoleName)
+	inheritedGuestRole, err := a.GetRoleByName(inheritedGuestRoleName)
 	if err != nil {
 		return nil, err
 	}
-	inheritedMemberPermissions := GetChannelModeratedPermissions(inheritedMemberRole)
-	inheritedGuestPermissions := GetChannelModeratedPermissions(inhertedGuestRole)
+
+	inheritedMemberPermissions := GetChannelModeratedPermissions(inheritedMemberRole.Permissions)
+	inheritedGuestPermissions := GetChannelModeratedPermissions(inheritedGuestRole.Permissions)
+
+	for _, moderationPatch := range channelModerationsPatch {
+		if moderationPatch.Roles.Members != nil && *moderationPatch.Roles.Members && !inheritedMemberPermissions[*moderationPatch.Name] {
+			return nil, &model.AppError{Message: "Cannot add a permission that is restricted by the team or system permission scheme"}
+		}
+		if moderationPatch.Roles.Guests != nil && *moderationPatch.Roles.Guests && !inheritedGuestPermissions[*moderationPatch.Name] {
+			return nil, &model.AppError{Message: "Cannot add a permission that is restricted by the team or system permission scheme"}
+		}
+	}
+
+	// Channel has no scheme so create one
+	if channel.SchemeId == nil || len(*channel.SchemeId) == 0 {
+		scheme := &model.Scheme{
+			Name:        model.NewId(),
+			DisplayName: model.NewId(),
+			Scope:       model.SCHEME_SCOPE_CHANNEL,
+		}
+		if scheme, err = a.CreateScheme(scheme); err != nil {
+			return nil, err
+		}
+		channel.SchemeId = &scheme.Id
+		a.UpdateChannelScheme(channel)
+	}
+
+	guestRoleName, memberRoleName, _, _ := a.GetSchemeRolesForChannel(channel.Id)
+	memberRole, err := a.GetRoleByName(memberRoleName)
+	if err != nil {
+		return nil, err
+	}
+
+	guestRole, err := a.GetRoleByName(guestRoleName)
+	if err != nil {
+		return nil, err
+	}
+
+	memberRolePatch := memberRole.RolePatchFromChannelModerationsPatch(channelModerationsPatch, "members")
+	guestRolePatch := guestRole.RolePatchFromChannelModerationsPatch(channelModerationsPatch, "guests")
```

Comment:
This block creates `RolePatchs` from a list of channel moderation patches for a each role (guest and member)